cmd/snap-bootstrap/initramfs_systemd_mount.go

```diff
@@ -45,6 +46,140 @@ Wants=%[1]s
 	doSystemdMount = doSystemdMountImpl
 )
 
+// forbiddenChars is a list of characters that are not allowed in any mount paths used in systemd-mount.
+const forbiddenChars = `\,:" `
+
+type mountOptions interface {
+	AppendOptions([]string) ([]string, error)
+}
+
+// overlayFsOptions groups the options to systemd-mount related to overlayfs.
+type overlayFsOptions struct {
+	// Directories to be used as lower layers of an overlay mount.
+	// It does not need to be on a writable filesystem.
+	LowerDirs []string
+	// Optional. A directory to be used as the upper layer of an overlay mount.
+	// This is normally on a writable filesystem.
+	UpperDir string
+	// Optional. A directory to be used as the workdir of an overlay mount.
+	// This needs to be an empty directory on the same filesystem as upperdir.
+	WorkDir string
+}
+
+// validate is used to perform consistency checks on the options related to overlayfs mounts.
+func (o *overlayFsOptions) validate() error {
+	if len(o.LowerDirs) <= 0 {
+		return errors.New("missing arguments for overlayfs mount: at least one lowerdir is required")
+	}
+
+	if len(o.UpperDir) > 0 && len(o.WorkDir) <= 0 {
+		return errors.New("an upperdir for an overlayfs mount was specified but workdir is missing")
+	}
+
+	if len(o.WorkDir) > 0 && len(o.UpperDir) <= 0 {
+		return errors.New("a workdir for an overlayfs mount was specified but upperdir is missing")
+	}
+
+	if strings.ContainsAny(o.UpperDir, forbiddenChars) {
+		return fmt.Errorf("upperdir overlayfs mount option contains forbidden characters. %q contains one of %q", o.UpperDir, forbiddenChars)
+	}
+	if strings.ContainsAny(o.WorkDir, forbiddenChars) {
+		return fmt.Errorf("workdir overlayfs mount option contains forbidden characters. %q contains one of %q", o.WorkDir, forbiddenChars)
+	}
+
```

Comment:
Should we move the validation of lowerdirs here? I don't see why it has to be separate. Once you do that the logic for assembling options to mount is not different, only that the checks are earlier.

Reply:
I'm a bit plus one on this, I don't see the consistency in validation currently.

I'd like to see either one `validate()` that calls sub-functions like `validateLowerDirs() / validateUpperDirs()` ... or put all the validation logic here

Reply:
I am not sure I understand the suggestion here, could you clarify a bit?

So having `validateLowerDirs()` also construct the lowerDirs string is indeed somewhat confusing but it avoids having to go through that list twice, which I thought was a nice property to keep mostly because I wouldn't have to have the same looping code in two places. So are you suggesting to do a loop once in `validate()` just for the validation and then do the same loop again in `AppendOptions()` to construct the `lowerDirs` string?

Reply:
That would be my opinion yes, unless there is a reasonable performance impact in doing so, then I think that it would produce cleaner code in this case with clearer intent. You can assume that the content is "safe" in the function that actually constructs the lowerDirs string.

I would also accept just renaming `validateLowerDirs` to something else, so it doesn't look like it's primary purpose is validation, if you are strongly opposed to my suggestion

Reply:
I made it that way. I actually pulled the validation logic in `validate` and the rest I just put directly in `AppendOptions` and removed the extra function. Thanks for the suggestions!